File: leetcode/2463-minimum-recolors-to-get-k-consecutive-black-blocks/minimum-recolors-to-get-k-consecutive-black-blocks.py
# Amir Mukhtarov, mukhtarov.amir.a@gmail.com


class Solution:
    def minimumRecolors(self, blocks: str, k: int) -> int:
        res = k
        recolor = 0

        left = 0

        for right in range(len(blocks)):
            if blocks[right] == 'W':
                recolor += 1
            if right - left + 1 == k:
                res = min(res, recolor)
                if blocks[left] == 'W':
                    recolor -= 1
                left += 1

        return res


if __name__ == '__main__':
    blocks = "WBBWWBBWBW"
    k = 7

    blocks = "WBWBBBW"
    k = 2

    s = Solution()
    print(s.minimumRecolors(blocks, k))

File: leetcode/2463-minimum-recolors-to-get-k-consecutive-black-blocks/minimum-recolors-to-get-k-consecutive-black-blocks.cpp
// Amir Mukhtarov, mukhtarov.amir.a@gmail.com
#include <iostream> 
#include <vector> 
#include <string> 
#include <algorithm> 
#include <unordered_map> 
#include <map> 
#include <unordered_set> 
#include <set> 
#include <queue> 
#include <stack> 
#include <deque> 
#include <cmath> 
#include <limits> 

using namespace std; 
 
class Solution {
public:
    int minimumRecolors(string blocks, int k) {
      int res = k;
      int recolor = 0;

      int left = 0;

      for (int right = 0; right <= blocks.size(); right++) {
        if (blocks[right] == 'W') {
          recolor++;
        }
        if (right - left + 1 == k) {
          res = min(res, recolor);
          if (blocks[left] == 'W') {
            recolor--;
          }
          left++;
        }
      }


      return res;
        
    }
};


int main(){ 
  string blocks = "WBBWWBBWBW";
  int k  = 7;

  Solution s = Solution();

  cout << s.minimumRecolors(blocks, k) << endl;

  return 0;
}
